### PR TITLE
RavenDB-8832: Implement and use PeepingTomStream

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

File: src/Raven.Client/Documents/Changes/DatabaseChanges.cs [hunk 1]
```diff
@@ -417,23 +418,24 @@ private DatabaseConnectionState GetOrAddConnectionState(string name, string watc
                     using (var parser = new UnmanagedJsonParser(context, state, "changes/receive"))
                     using (var builder = new BlittableJsonDocumentBuilder(context, BlittableJsonDocumentBuilder.UsageMode.None, "readArray/singleResult", parser, state))
                     {
-                        if (await UnmanagedJsonParserHelper.ReadAsync(stream, parser, state, buffer).ConfigureAwait(false) == false)
+                        var peepingTomStream = new PeepingTomStream(stream, context);
```
Comment (ayende) [inline src/Raven.Client/Documents/Changes/DatabaseChanges.cs]: We should be disposing this

File: src/Raven.Client/Documents/Commands/MultiGet/MultiGetCommand.cs [hunk 2]
```diff
@@ -119,19 +120,20 @@ public override void SetResponseRaw(HttpResponseMessage response, Stream stream,
             using (var parser = new UnmanagedJsonParser(context, state, "multi_get/response"))
             using (context.GetManagedBuffer(out JsonOperationContext.ManagedPinnedBuffer buffer))
             {
-                if (UnmanagedJsonParserHelper.Read(stream, parser, state, buffer) == false)
-                    ThrowInvalidResponse();
+                var peepingTomStream = new PeepingTomStream(stream, context);
```
Comment (ayende) [inline src/Raven.Client/Documents/Commands/MultiGet/MultiGetCommand.cs]: Need to dispose this

File: src/Raven.Client/Documents/Session/Operations/StreamOperation.cs [hunk 3]
```diff
@@ -84,63 +85,65 @@ public IEnumerator<BlittableJsonReaderObject> SetResult(StreamResult response)
             using (var parser = new UnmanagedJsonParser(_session.Context, state, "stream contents"))
             using (_session.Context.GetManagedBuffer(out buffer))
             {
-                if (UnmanagedJsonParserHelper.Read(response.Stream, parser, state, buffer) == false)
-                    UnmanagedJsonParserHelper.ThrowInvalidJson();
+                var peepingTomStream = new PeepingTomStream(response.Stream, _session.Context);
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/Operations/StreamOperation.cs]: Need to dispose this

File: src/Raven.Client/Documents/Session/Operations/StreamOperation.cs [hunk 4]
```diff
@@ -84,63 +85,65 @@ public IEnumerator<BlittableJsonReaderObject> SetResult(StreamResult response)
...
                 if (state.CurrentTokenType != JsonParserToken.EndObject)
-                    UnmanagedJsonParserHelper.ThrowInvalidJson();
+                    UnmanagedJsonParserHelper.ThrowInvalidJson(peepingTomStream);
             }
         }
 
         private static void HandleStreamQueryStats(JsonOperationContext context ,StreamResult response, UnmanagedJsonParser parser, JsonParserState state, JsonOperationContext.ManagedPinnedBuffer buffer, StreamQueryStatistics streamQueryStatistics = null)
         {
-            var property = UnmanagedJsonParserHelper.ReadString(context, response.Stream, parser, state, buffer);
+            var peepingTomStream = new PeepingTomStream(response.Stream, context);
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/Operations/StreamOperation.cs]: Dispose this

File: src/Raven.Client/Documents/Session/Operations/StreamOperation.cs [hunk 5]
```diff
@@ -167,6 +170,7 @@ private class YieldStreamResults : IAsyncEnumerator<BlittableJsonReaderObject>
             public YieldStreamResults(InMemoryDocumentSessionOperations session, StreamResult stream, bool isQueryStream)
             {
                 _response = stream;
+                _peepingTomStream = new PeepingTomStream(_response.Stream, session.Context);
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/Operations/StreamOperation.cs:175]: dispose this

File: src/Raven.Server/Smuggler/Documents/StreamSource.cs [hunk 6]
```diff
@@ -42,7 +40,7 @@ public class StreamSource : ISmugglerSource
 
         public StreamSource(Stream stream, DocumentsOperationContext context)
         {
-            _stream = stream;
+            _peepingTomStream = new PeepingTomStream(stream, context);
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/StreamSource.cs:43]: dispose this

File: src/Sparrow/PeepingTomStream.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                start = 0;
+                size = _pos;
+            }
+            else
+            {
+                start = _pos;
+                size = BufferWindowSize;
+            }
+            while ((_bufferWindow.Buffer.Array[start] & 0x80) != 0)
```
Comment (ayende) [inline src/Sparrow/PeepingTomStream.cs]: Explain why we are doing this

File: src/Sparrow/PeepingTomStream.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                    Debug.Assert(nextLength <= BufferWindowSize);
+                    Memory.Copy(pDest, pBufferStart + newTotal, nextLength);
+                }
+                else
+                {
+                    Memory.Copy(pBufferWindowStart, pBufferStart, totalToRead);
+                }
+            }
+
+            _pos += totalToRead;
```
Comment (ayende) [inline src/Sparrow/PeepingTomStream.cs]: Maybe do this inside the previous if, so it is all in a single place?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Session/Operations/StreamOperation.cs (hunk 5 above)
Comment (aviviadi) [reply src/Raven.Client/Documents/Session/Operations/StreamOperation.cs:175]: now disposed by containing class

File: src/Raven.Server/Smuggler/Documents/StreamSource.cs (hunk 6 above)
Comment (aviviadi) [reply src/Raven.Server/Smuggler/Documents/StreamSource.cs:43]: now disposed by containing class

**Pull request merged**